### ISSUE TITLE
Linking error for enumeration facade

### ISSUE THREAD

**Issue by lastsys** [scala-js/scala-js#2244]
Reproduction of issue is given in https://github.com/lastsys/scala-js-issue2244. Just follow the instructions to obtain the error message.

The following code gives `Cannot access module for non-module` error during `fastOptJS` when accessing one of the enumerations below. Could be a problem where the `type` is mixed up with the `object`?

package.scala:
```
package org.scalajs.openui5.sap.ui

package object ux3 {
  type NotificationBarStatus = String
}
```

NotificationBarStatus.scala:
```
package org.scalajs.openui5.sap.ui.ux3

import scala.scalajs.js
import scala.scalajs.js.annotation.JSName

@JSName("sap.ui.ux3.NotificationBarStatus")
@js.native
object NotificationBarStatus extends js.Object {
  val Default: NotificationBarStatus = js.native
  val Max: NotificationBarStatus = js.native
  val Min: NotificationBarStatus = js.native
  val None: NotificationBarStatus = js.native
}
```

**Comment by lastsys** [comment]
Reproduction of issue is given in https://github.com/lastsys/scala-js-issue2244.

**Comment by sjrd** [comment]
Some progress status: this is a scalac bug (2, in fact!)

Declaring a `type` in the package object, and the "companion" `object` in a different file, is definitely not allowed. If the type is a `trait` instead, scalac complains with:
```
[error] /localhome/doeraene/projects/scalajs-openui5/src/main/scala/org/scalajs/openui5/sap/ui/ux3/package.scala:7: Companions 'trait NotificationBarStatus' and 'object NotificationBarStatus' must be defined in same file:
[error]   Found in /localhome/doeraene/projects/scalajs-openui5/src/main/scala/org/scalajs/openui5/sap/ui/ux3/package.scala and /localhome/doeraene/projects/scalajs-openui5/src/main/scala/org/scalajs/openui5/sap/ui/ux3/NotificationBarStatus.scala
[error]   trait NotificationBarStatus
[error]         ^
```
It seems scalac has a bug in the case of a type alias: it should report a compile error, but it doesn't. The compilation results in broken .class files, where the `object` is basically overridden by an empty, synthetic object. This causes `fastOptJS` to go berserk, of course, and rightfully so.

So, OK, let's fix that, and move the `object` inside the package object:
```scala
package object ux3 {
  type NotificationBarStatus = String

  @JSName("sap.ui.ux3.NotificationBarStatus")
  @js.native
  object NotificationBarStatus extends js.Object {
    val Default: NotificationBarStatus = js.native
    val Max: NotificationBarStatus = js.native
    val Min: NotificationBarStatus = js.native
    val None: NotificationBarStatus = js.native
  }
}
```
Now we get this compile error:
```
[error] /localhome/doeraene/projects/scalajs-openui5/src/main/scala/org/scalajs/openui5/sap/ui/ux3/package.scala:11: NotificationBarStatus extends scala.Serializable which does not extend js.Any.
[error]   object NotificationBarStatus extends js.Object {
[error]          ^
```
Wait, what!? `NotificationBarStatus` does not extend `Serializable`. scalac, what are you babbling about?

Hum, let's change the `type` alias so it's an abstract type instead
```scala
type NotificationBarStatus
```
Oh! Now it compiles. What!?

If we *upper-bound* the type with `<: String`, the error comes back. If we *lower-bound* with `>: String` instead, the error disappears. If we turn the type into a `trait`, it's fine. But if we make the trait `extends Serializable`, the error comes back.

OK, so, scalac, are you adding `extends Serializable` to my `object` if its "companion" is somehow known to extend `Serializable`? Even when it's an abstract, upper-bounded type alias? Come on, that must be a bug too, right?

We can reproduce that behavior without Scala.js:
```scala
package bug

package object testing {
  type Foo = String

  object Foo
}

object Test {
  def main(args: Array[String]): Unit = {
    // This should not display scala.Serializable
    println(testing.Foo.getClass.getInterfaces().mkString(", "))

    // This should not even compile:
    println(testing.Foo: Serializable)
  }
}
```
```
scalac-bug-serializable-companion$ ~/opt/scala-2.11.2/bin/scalac Test.scala
scalac-bug-serializable-companion$ ~/opt/scala-2.11.2/bin/scala -cp . bug.Test
interface scala.Serializable
```

**Comment by sjrd** [comment]
Second bug reported: https://issues.scala-lang.org/browse/SI-9668

**Comment by sjrd** [comment]
So, I'm afraid that's a won't fix as far as Scala.js is concerned. The bug has to be fixed on scalac's side.

**Comment by sjrd** [comment]
FYI: This has been fixed in Scala 2.12.2. See https://github.com/scala/scala/pull/5602
